app/master/build_artifact.py

```diff
@@ -103,3 +110,28 @@ def _write_timing_data_to_file(self, path, timing_data):
         :type timing_data: dict[str, float]
         """
         app.util.fs.write_file(json.dumps(timing_data), path)
+
+    @staticmethod
+    def artifact_directory(build_id, subjob_id=None, atom_id=None, result_root=None):
+        """
+        To get the full path to an atom artifact id, the caller must specify all id's: build_id, subjob_id, atom_id.
+        To get the path to just the build artifact directory, the caller must only specify the build_id.
+
+        If the caller specifies exactly one of the subjob_id or atom_id, it is a fatal error, and this method
```

Comment:
How about splitting the method into two since looks like there're two use cases. Of course, underneath, they can share the same private method for implementation.